File: 04. The 3n + 1 problem/src/Solution.py
class Solution():
    def __init__(self) -> None:
        # 利用dict儲存已經出現過的狀態，初始化1為1步
        self.table = {1 : 1}
        self.main()
    
    def main(self) -> None:
        #資料讀取
        import sys
        input = sys.stdin.read
        data = list(map(int, input().split()))

        data_length = len(data)

        for index in range(0, data_length, 2):
            start = min(data[index], data[index + 1])
            end = max(data[index], data[index + 1])
            maxmium = 1

            for number in range(start, end + 1):
                
                maxmium = max(maxmium, self.process(number))

            print(f"{data[index]} {data[index + 1]} {maxmium}")

    
    # 計算number cycle length
    def process(self, number : int) -> int:
        if number in self.table:
            return self.table[number]
        if number % 2 == 1:
            self.table[number] = self.process(number * 3 + 1) + 1
        else:
            self.table[number] = self.process(number // 2) + 1
        return self.table[number]
    
if __name__ == "__main__":
    run = Solution()

File: 04. The 3n + 1 problem/src/Solution.cpp
#include <iostream>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    Solution() {
        table[1] = 1;
        this->main();
    }

private:
    unordered_map<int, int> table;

    void main() {
        int a, b;

        while (cin >> a >> b) {
            int i = min(a, b);
            int j = max(a, b);

            int maxCycleLength = 0;

            for (int k = i; k <= j; ++k) {
                maxCycleLength = max(maxCycleLength, process(k));
            }

            cout << a << ' ' << b << ' ' << maxCycleLength << endl;
        }
    }

    int process(int number) {
        if (table.find(number) != table.end()) {
            return table[number];
        }

        int result;
        if (number % 2 == 0) {
            result = process(number / 2) + 1;
        }
        else {
            result = process(number * 3 + 1) + 1;
        }
        table[number] = result;
        return result;
    }
};

int main() {
    Solution();
    return 0;
}
